src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -0,0 +1,282 @@
+        Lazy<IProjectHotReloadNotificationService> projectHotReloadNotificationService,
+        IProjectHotReloadLaunchProvider launchProvider,
+        IProjectHotReloadBuildManager buildManager,
+        IProjectHotReloadAgent hotReloadAgent)
+        : base(threadingService.JoinableTaskContext)
+    {
+        _configuredProject = project;
+        _unconfiguredProject = project.UnconfiguredProject;
+        _threadingService = threadingService;
+        _hotReloadDiagnosticOutputService = hotReloadDiagnosticOutputService;
+        _projectHotReloadNotificationService = projectHotReloadNotificationService;
+        _hotReloadAgent = hotReloadAgent;
+        _semaphore = ReentrantSemaphore.Create(
+            initialCount: 1,
+            joinableTaskContext: threadingService.JoinableTaskContext.Context,
+            mode: ReentrantSemaphore.ReentrancyMode.Freeform);
+    }
+
+    public Task<bool> TryCreatePendingSessionAsync(
+        IProjectHotReloadLaunchProvider launchProvider,
+        IDictionary<string, string> environmentVariables,
+        DebugLaunchOptions launchOptions,
+        ILaunchProfile launchProfile)
+    {
+        return _semaphore.ExecuteAsync(TryCreatePendingSessionInternalAsync).AsTask();
+
+        async ValueTask<bool> TryCreatePendingSessionInternalAsync()
+        {
+            if (await ProjectSupportsHotReloadAsync())
+            {
+                if (await ProjectSupportsStartupHooksAsync())
+                {
+                    string name = Path.GetFileNameWithoutExtension(_unconfiguredProject.FullPath);
+                    HotReloadSessionState hotReloadSessionState = new((HotReloadSessionState sessionState) =>
+                    {
+                        int count;
+                        lock (_activeSessionStates)
+                        {
+                            _activeSessionStates.Remove(sessionState);
+                            count = _activeSessionStates.Count;
+                        }
+
+                        if (count == 0)
+                        {
+                            _threadingService.ExecuteSynchronously(() => _projectHotReloadNotificationService.Value.SetHotReloadStateAsync(isInHotReload: false));
+                        }
+                    }, _threadingService);
+
+                    IProjectHotReloadSession projectHotReloadSession = _hotReloadAgent.CreateHotReloadSession(
+                        name: name,
+                        id: _nextUniqueId++,
+                        callback: hotReloadSessionState,
+                        launchProfile: launchProfile,
+                        configuredProject: _configuredProject,
+                        debugLaunchOptions: launchOptions);
+
+                    hotReloadSessionState.Session = projectHotReloadSession;
+                    await projectHotReloadSession.ApplyLaunchVariablesAsync(environmentVariables, default);
+
+                    _pendingSessionState = hotReloadSessionState;
```

Comment:
To ensure "pending" data is not lost, consider validating that this is null before storing a new value here.


```suggestion
                    Assumes.Null(_pendingSessionState, "Attempt to overwrite unprocessed pending hot reload sessions state.");
                    _pendingSessionState = hotReloadSessionState;
```